prom_sky_con/pkg/store_prom.go

```diff
@@ -0,0 +1,204 @@
+	"github.com/skydive-project/skydive/flow"
+	"github.com/skydive-project/skydive/logging"
+)
+
+// Connections with no traffic for timeout period (in seconds) are no longer reported
+const defaultConnectionTimeout = 60
+
+// Run the cleanup process every cleanupTime seconds
+const cleanupTime = 120
+
+// data to be exported to prometheus
+// one data item per tuple
+var (
+	bytesSent = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "skydive_network_connection_total_bytes",
+			Help: "Number of bytes that have been transmmitted on this connection",
+		},
+		[]string{"initiator_ip", "target_ip", "initiator_port", "target_port", "direction", "node_tid"},
+	)
+)
+
+type storePrometheus struct {
+	pipeline          *core.Pipeline // not currently used
+	port              string
+	connectionCache   *cache.Cache
+	connectionTimeout int64
+}
+
+// we maintain a cache to keep track of connections that have been active/inactive
+type cacheEntry struct {
+	label1    prometheus.Labels
+	label2    prometheus.Labels
+	timeStamp int64
+}
+
+// SetPipeline setup; called by core/pipeline.NewPipeline
+func (s *storePrometheus) SetPipeline(pipeline *core.Pipeline) {
+	s.pipeline = pipeline
+}
+
+// StoreFlows store flows info in memory, before being shipped out
+// For each flow reported, prepare prometheus entries, one for each direction of data flow.
+func (s *storePrometheus) StoreFlows(flows map[core.Tag][]interface{}) error {
+	secs := time.Now().Unix()
+	for _, val := range flows {
+		for _, i := range val {
+			f := i.(*flow.Flow)
+			if f.Transport == nil {
+				continue
+			}
+			initiator_ip := f.Network.A
+			target_ip := f.Network.B
+			initiator_port := strconv.FormatInt(f.Transport.A, 10)
+			target_port := strconv.FormatInt(f.Transport.B, 10)
+			node_tid := f.NodeTID
+			label1 := prometheus.Labels{
+				"initiator_ip":   initiator_ip,
+				"target_ip":      target_ip,
+				"initiator_port": initiator_port,
+				"target_port":    target_port,
+				"direction":      "initiator_to_target",
+				"node_tid":       node_tid,
+			}
+			label2 := prometheus.Labels{
+				"initiator_ip":   initiator_ip,
+				"target_ip":      target_ip,
+				"initiator_port": initiator_port,
+				"target_port":    target_port,
+				"direction":      "target_to_initiator",
+				"node_tid":       node_tid,
+			}
+			// post the info to prometheus
+			bytesSent.With(label1).Set(float64(f.Metric.ABBytes))
+			bytesSent.With(label2).Set(float64(f.Metric.BABytes))
+			cEntry := cacheEntry{
+				label1:    label1,
+				label2:    label2,
+				timeStamp: secs,
+			}
+			s.connectionCache.Set(f.L3TrackingID, cEntry, 0)
+		}
+	}
+	return nil
+}
+
+// cleanupExpiredEntries - any entry that has expired should be removed from the prometheus reporting and cache
+func (s *storePrometheus) cleanupExpiredEntries() {
+	secs := time.Now().Unix()
+	expireTime := secs - s.connectionTimeout
+	entriesMap := s.connectionCache.Items()
+	for k, v := range entriesMap {
```

Comment:
o(n) ... this is very expensive ... better if we can reduce to the order of entries which have expired rather than order of all the connection

Reply:
I agree, but I don't know how to do that. We get no traffic from the inactive connections, so we have no easy way to get a handle on them, unless we go through the entire list. The timeout period is every few minutes, so we don't do this often. Any concrete suggestions?

Reply:
use a list with the head pointing to the oldest connection. then each timeout period scan the connections from the head until you reach recent connections. this way you only scan the head of the list and not the entire list per each timeout interval.

Reply:
I use the cache to keep track of all the active connections. When a connection is quiet for too long, I stop reporting its statistics to prometheus. An LRU cache with evictions would remove the elements behind my back and I would miss the opportunity to cleanup my data structure. I guess I don't need a cache with all its overhead, but simply a map of all my connections. I would still have to go over the entire map to identify which items to cleanup, but that would be cheaper than going through the items in a full-fledged cache. What do you think?

Reply:
I added an LRU mechanism by pushing used elements to the end of a list each time they are updated. This leaves the elements in the list in LRU order, providing a simple way to delete the expired entries. Thanks for the suggestion.

Reply:
thanks!